Basic Skript API for AABs and Spheres

sovdeeth opened the pull request:
- [X] Create AAB
- [x] Create Sphere
- [x] AAB Dimensions
  - [X] get
  - [x] set
- [x] Sphere Dimensions
  - [x] get
  - [x] set 
- [x] bound center 
  - [x] get
  - [x] set
- [x] AAB min/max
  - [x] get
  - [x] set 
- [x] contains condition
- [x] intersects condition

sovdeeth committed:
shadow kotlin?

sovdeeth committed:
register bounding box type, default.lang

sovdeeth committed:
wrong class

sovdeeth committed:
remove extra user

sovdeeth committed:
is ti the plural that's screwing it up?

sovdeeth committed:
unprimitive the class reference

sovdeeth committed:
silly me, you don't need % there

sovdeeth committed:
Update ExprDimensions.kt

sovdeeth committed:
woo line endings!

sovdeeth committed:
Update build_and_test_action.yml

sovdeeth committed:
Update build_and_test_action.yml

sovdeeth committed:
Update build_and_test_action.yml

sovdeeth committed:
Better inheritance structure for bounds, dimension properties, Skript API for dimensions/creation

sovdeeth committed:
forgot the cuboid type

sovdeeth committed:
fix skript test typos

sovdeeth committed:
wrong expression order

sovdeeth committed:
hopefully the last typo

that srkipt-test gradle plugin can't come quick enough pickle!

Sparky200 commented on src/main/kotlin/com/sovdee/skriptbounds/bounds/AxisAlignedBox.kt:
```diff
@@ -81,13 +98,19 @@ data class AxisAlignedBox(val min: Vector3d, val max: Vector3d) : BoundingBox {
     override fun expandUniform(amount: Double) {
         min.add(-amount, -amount, -amount)
         max.add(amount, amount, amount)
```
You might be interested in using an operator function (as a receiver function on Vector3d) to have this syntax:
```suggestion
        max += amount
```
That one specifically would be:
```kt
operator fun Vector3d.plusAssign(amount: Int) {
    add(amount, amount, amount)
}
```

Sparky200 commented on src/main/kotlin/com/sovdee/skriptbounds/bukkit/BukkitConversion.kt:
```diff
@@ -13,7 +13,8 @@ fun toPosition(any: Any?) : Vector3d? {
-    return vector?.toVector3d()
+    val v = vector?.toVector3d()
+    return if (v?.isFinite == true) v else null
 }
```
This could be a receiver, though it's unfortunate that there's already a toVector3d() on that vector.
```kt
fun Vector?.toPosition() {
    val v = vector?.toVector3d()
    return if (v?.isFinite == true) v else null
}
```
To clarify, this can still be called on nullables

Sparky200 commented on src/main/kotlin/com/sovdee/skriptbounds/elements/Types.kt:
```diff
@@ -0,0 +1,115 @@
+                            fields.putPrimitive("centery", o.center.y)
+                            fields.putPrimitive("centerz", o.center.z)
+                            fields.putPrimitive("rsquared", o.radiusSquared)
+                            return fields
```
Maybe:
```suggestion
                            Fields().apply {
                                putPrimitive("centerx", o.center.x)
                                putPrimitive("centery", o.center.y)
                                putPrimitive("centerz", o.center.z)
                                putPrimitive("rsquared", o.radiusSquared)
                            }
```
I would suggest `with` syntax but it only returns the explicit final value of the block (so you would put `this` at the end of the block).
```kt
with (Fields()) {
    ...
    this
}

Sparky200 commented on src/main/kotlin/com/sovdee/skriptbounds/elements/properties/ExprDimensions.kt:
```diff
@@ -27,20 +36,85 @@ class ExprDimensions : SimplePropertyExpression<AxisAlignedBox, Double>() {
+                else -> null
+            }
+
+        assert(false)
```
Would it be better to raise an exception with a message here?
```suggestion
        throw IllegalArgumentException("Bounding box type '${from::class.name}' is not supported!")
```
Assertion does make sense, but this may provide better insight on what went wrong

Sparky200 reviewed (commented):
LGTM, I don't see any over-the-top misuse anywhere. Most of my comments are just another way to do something which is just style preference.

sovdeeth committed:
box center expression

sovdeeth committed:
pesky space

sovdeeth committed:
cond contains attempt 1

sovdeeth committed:
use SimpleExpression#check()

sovdeeth committed:
wrong isNegated logic

sovdeeth committed:
fix test error with AND on 50% success

sovdeeth committed:
ditto

sovdeeth committed:
many-many contains is just confusing.

"ABC" doesn't contain "A" or "E" -> false. I'm not a fan of this behavior but it matches skript contains.

sovdeeth committed:
cuboid corners and some syntax standardization

sovdeeth committed:
some silly oversights

sovdeeth committed:
Update ExprCuboidCorners.kt

sovdeeth committed:
more standardization

sovdeeth committed:
fix logic in fixMinMax()

sovdeeth committed:
this one's a bit embarassing

sovdeeth committed:
Experimenting with jetbrains ai docs generation

saves some time but certainly isn't perfect.

sovdeeth committed:
bound intersection
